File: python/weather_estimator.py
"""Deterministic GFS fair-value source for supported temperature markets."""

from __future__ import annotations

import math
import re
import time
from dataclasses import dataclass
from datetime import date, datetime, timedelta
from typing import Callable, Optional

import requests

from models import MarketInfo

OPEN_METEO_ENSEMBLE_URL = "https://ensemble-api.open-meteo.com/v1/ensemble"
PROVIDER_NAME = "weather_gfs"
MODEL_NAME = "gfs_seamless"

STATIONS = {
    "KLGA": (40.7769, -73.8740),
    "KNYC": (40.7794, -73.9692),
    "KORD": (41.9742, -87.9073),
    "KMIA": (25.7959, -80.2870),
    "KLAX": (33.9425, -118.4081),
    "KDEN": (39.8617, -104.6731),
}

CITY_STATIONS = (
    (r"\bnew york city\b|\bnew york\b|\bnyc\b", "KLGA"),
    (r"\bchicago\b", "KORD"),
    (r"\bmiami\b", "KMIA"),
    (r"\blos angeles\b", "KLAX"),
    (r"\bdenver\b", "KDEN"),
)

MONTHS = {
    "january": 1, "jan": 1, "february": 2, "feb": 2, "march": 3, "mar": 3,
    "april": 4, "apr": 4, "may": 5, "june": 6, "jun": 6, "july": 7, "jul": 7,
    "august": 8, "aug": 8, "september": 9, "sep": 9, "october": 10, "oct": 10,
    "november": 11, "nov": 11, "december": 12, "dec": 12,
}


@dataclass(frozen=True)
class WeatherMarketSpec:
    station: str
    latitude: float
    longitude: float
    target_date: date
    metric: str
    lower_f: Optional[float]
    upper_f: Optional[float]


@dataclass(frozen=True)
class WeatherEstimate:
    probability: float
    member_count: int
    matching_members: int
    station: str
    reasoning: str


def _fallback_year(market: MarketInfo) -> int:
    if market.end_date:
        try:
            return datetime.fromisoformat(market.end_date.replace("Z", "+00:00")).year
        except ValueError:
            pass
    return date.today().year


def _extract_date(text: str, year: int) -> Optional[date]:
    month_names = "|".join(sorted(MONTHS, key=len, reverse=True))
    match = re.search(rf"\b({month_names})\s+(\d{{1,2}})(?:\s*,?\s*(\d{{4}}))?\b", text, re.I)
    if match:
        try:
            return date(int(match.group(3) or year), MONTHS[match.group(1).lower()], int(match.group(2)))
        except ValueError:
            return None
    return None


def _station_for_market(market: MarketInfo, text: str) -> Optional[str]:
    explicit = re.search(r"(?:[?&]site=|\bstation\s+)([a-z0-9]{4})\b", market.description or "", re.I)
    if explicit:
        station = explicit.group(1).upper()
        return station if station in STATIONS else None
    for pattern, station in CITY_STATIONS:
        if re.search(pattern, text, re.I):
            return station
    return None


def _temperature_range(text: str) -> tuple[Optional[float], Optional[float]] | None:
    number = r"(-?\d+(?:\.\d+)?)"
    bracket = re.search(rf"{number}\s*(?:-|–|—|to)\s*{number}\s*°?\s*f\b", text, re.I)
    if bracket:
        lower, upper = float(bracket.group(1)), float(bracket.group(2))
        return min(lower, upper), max(lower, upper)
    tail = re.search(rf"{number}\s*°?\s*f\s*(?:or\s+)?(below|lower|less|under|higher|above|more|over)\b", text, re.I)
    if tail:
        threshold = float(tail.group(1))
        direction = tail.group(2).lower()
        return (None, threshold) if direction in {"below", "lower", "less", "under"} else (threshold, None)
    return None


def parse_weather_market(market: MarketInfo) -> Optional[WeatherMarketSpec]:
    text = f"{market.question} {market.event_title} {market.slug}".lower()
    if market.category.lower() != "weather" and "temperature" not in text:
        return None
    if re.search(r"\b(highest|high temperature|daily high)\b", text):
        metric = "high"
    elif re.search(r"\b(lowest|low temperature|daily low)\b", text):
        metric = "low"
    else:
        return None
    bounds = _temperature_range(text)
    station = _station_for_market(market, text)
    target_date = _extract_date(text, _fallback_year(market))
    if bounds is None or station is None or target_date is None:
        return None
    latitude, longitude = STATIONS[station]
    return WeatherMarketSpec(station, latitude, longitude, target_date, metric, bounds[0], bounds[1])


def probability_for_range(members: list[float], lower_f: Optional[float], upper_f: Optional[float]) -> float:
    if not members:
        return 0.0
    lower_edge = -math.inf if lower_f is None else lower_f - 0.5
    upper_edge = math.inf if upper_f is None else upper_f + 0.5
    return sum(lower_edge <= value < upper_edge for value in members) / len(members)


class WeatherEstimator:
    def __init__(self, get: Callable = requests.get):
        self._get = get
        self._cache: dict[tuple[float, float, str], tuple[float, dict]] = {}

    def estimate(self, market: MarketInfo) -> Optional[WeatherEstimate]:
        spec = parse_weather_market(market)
        if spec is None:
            return None
        today = date.today()
        if spec.target_date < today or spec.target_date > today + timedelta(days=16):
            return None
        data = self._forecast(spec)
        if data is None:
            return None
        prefix = "temperature_2m_max" if spec.metric == "high" else "temperature_2m_min"
        daily = data.get("daily", {})
        members: list[float] = []
        for key, values in daily.items():
            if not key.startswith(prefix) or not isinstance(values, list) or not values:
                continue
            try:
                value = float(values[0])
            except (TypeError, ValueError):
                continue
            if math.isfinite(value):
                members.append(value)
        if len(members) < 10:
            return None
        raw_probability = probability_for_range(members, spec.lower_f, spec.upper_f)
        probability = min(0.95, max(0.05, raw_probability))
        matching = round(raw_probability * len(members))
        return WeatherEstimate(probability, len(members), matching, spec.station, "GFS ensemble")

    def _forecast(self, spec: WeatherMarketSpec) -> Optional[dict]:
        cache_key = (spec.latitude, spec.longitude, spec.target_date.isoformat())
        cached = self._cache.get(cache_key)
        if cached and time.monotonic() - cached[0] < 900:
            return cached[1]
        try:
            response = self._get(
                OPEN_METEO_ENSEMBLE_URL,
                params={"latitude": spec.latitude, "longitude": spec.longitude,
                        "daily": "temperature_2m_max,temperature_2m_min",
                        "temperature_unit": "fahrenheit", "timezone": "auto",
                        "start_date": spec.target_date.isoformat(),
                        "end_date": spec.target_date.isoformat(), "models": MODEL_NAME},
                timeout=15,
            )
            response.raise_for_status()
            data = response.json()
        except (requests.RequestException, ValueError, TypeError):
            return None
        self._cache[cache_key] = (time.monotonic(), data)
        return data

File: dotnet/PolymarketBot/Services/WeatherEstimator.cs
using System.Globalization;
using System.Text.Json;
using System.Text.RegularExpressions;
using PolymarketBot.Models;

namespace PolymarketBot.Services;

public sealed record WeatherEstimateResult(double Probability, int MemberCount, int MatchingMembers, string Station, string Reasoning);

public sealed class WeatherEstimator(HttpClient http)
{
    public const string ProviderName = "weather_gfs";
    public const string ModelName = "gfs_seamless";
    private const string Endpoint = "https://ensemble-api.open-meteo.com/v1/ensemble";
    private static readonly Dictionary<string, (double Lat, double Lon)> Stations = new(StringComparer.OrdinalIgnoreCase)
    {
        ["KLGA"] = (40.7769, -73.8740), ["KNYC"] = (40.7794, -73.9692),
        ["KORD"] = (41.9742, -87.9073), ["KMIA"] = (25.7959, -80.2870),
        ["KLAX"] = (33.9425, -118.4081), ["KDEN"] = (39.8617, -104.6731),
    };
    private readonly Dictionary<string, (DateTimeOffset At, Dictionary<string, double[]> Data)> _cache = [];
    private sealed record Spec(string Station, double Lat, double Lon, DateOnly Date, string Metric, double? Lower, double? Upper);

    public static double ProbabilityForRange(IEnumerable<double> members, double? lower, double? upper)
    {
        var values = members.ToList();
        if (values.Count == 0) return 0;
        var low = lower.HasValue ? lower.Value - .5 : double.NegativeInfinity;
        var high = upper.HasValue ? upper.Value + .5 : double.PositiveInfinity;
        return (double)values.Count(value => value >= low && value < high) / values.Count;
    }

    private static Spec? Parse(MarketInfo market)
    {
        var text = $"{market.Question} {market.EventTitle} {market.Slug}";
        if (!market.Category.Equals("weather", StringComparison.OrdinalIgnoreCase) &&
            !text.Contains("temperature", StringComparison.OrdinalIgnoreCase)) return null;
        var metric = Regex.IsMatch(text, @"\b(highest|high temperature|daily high)\b", RegexOptions.IgnoreCase)
            ? "high" : Regex.IsMatch(text, @"\b(lowest|low temperature|daily low)\b", RegexOptions.IgnoreCase) ? "low" : null;
        if (metric is null) return null;
        var stationMatch = Regex.Match(market.Description ?? "", @"(?:[?&]site=|\bstation\s+)([a-z0-9]{4})\b", RegexOptions.IgnoreCase);
        var station = stationMatch.Success ? stationMatch.Groups[1].Value.ToUpperInvariant() : CityStation(text);
        if (station is null || !Stations.TryGetValue(station, out var coords)) return null;
        var range = TemperatureRange(text);
        var targetDate = TargetDate(text, market.EndDate);
        return range is null || targetDate is null ? null
            : new Spec(station, coords.Lat, coords.Lon, targetDate.Value, metric, range.Value.Lower, range.Value.Upper);
    }

    private static string? CityStation(string text)
    {
        if (Regex.IsMatch(text, @"\b(new york city|new york|nyc)\b", RegexOptions.IgnoreCase)) return "KLGA";
        if (Regex.IsMatch(text, @"\bchicago\b", RegexOptions.IgnoreCase)) return "KORD";
        if (Regex.IsMatch(text, @"\bmiami\b", RegexOptions.IgnoreCase)) return "KMIA";
        if (Regex.IsMatch(text, @"\blos angeles\b", RegexOptions.IgnoreCase)) return "KLAX";
        if (Regex.IsMatch(text, @"\bdenver\b", RegexOptions.IgnoreCase)) return "KDEN";
        return null;
    }

    private static (double? Lower, double? Upper)? TemperatureRange(string text)
    {
        var bracket = Regex.Match(text, @"(-?\d+(?:\.\d+)?)\s*(?:-|–|—|to)\s*(-?\d+(?:\.\d+)?)\s*°?\s*f\b", RegexOptions.IgnoreCase);
        if (bracket.Success)
        {
            var a = double.Parse(bracket.Groups[1].Value, CultureInfo.InvariantCulture);
            var b = double.Parse(bracket.Groups[2].Value, CultureInfo.InvariantCulture);
            return (Math.Min(a, b), Math.Max(a, b));
        }
        var tail = Regex.Match(text, @"(-?\d+(?:\.\d+)?)\s*°?\s*f\s*(?:or\s+)?(below|lower|less|under|higher|above|more|over)\b", RegexOptions.IgnoreCase);
        if (!tail.Success) return null;
        var threshold = double.Parse(tail.Groups[1].Value, CultureInfo.InvariantCulture);
        var direction = tail.Groups[2].Value.ToLowerInvariant();
        return direction is "below" or "lower" or "less" or "under" ? (null, threshold) : (threshold, null);
    }

    private static DateOnly? TargetDate(string text, string endDate)
    {
        var year = DateTimeOffset.TryParse(endDate, out var end) ? end.Year : DateTime.UtcNow.Year;
        var match = Regex.Match(text,
            @"\b(January|February|March|April|May|June|July|August|September|October|November|December)\s+(\d{1,2})\b",
            RegexOptions.IgnoreCase);
        if (!match.Success) return null;
        var month = DateTime.ParseExact(match.Groups[1].Value, "MMMM", CultureInfo.InvariantCulture).Month;
        return new DateOnly(year, month, int.Parse(match.Groups[2].Value, CultureInfo.InvariantCulture));
    }

    public async Task<WeatherEstimateResult?> EstimateAsync(MarketInfo market, CancellationToken ct = default)
    {
        var spec = Parse(market);
        if (spec is null) return null;
        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        if (spec.Date < today || spec.Date > today.AddDays(16)) return null;
        var forecast = await ForecastAsync(spec, ct);
        if (forecast is null) return null;
        var prefix = spec.Metric == "high" ? "temperature_2m_max" : "temperature_2m_min";
        var members = forecast.Where(item => item.Key.StartsWith(prefix, StringComparison.Ordinal))
            .SelectMany(item => item.Value.Take(1)).Where(double.IsFinite).ToList();
        if (members.Count < 10) return null;
        var raw = ProbabilityForRange(members, spec.Lower, spec.Upper);
        var probability = Math.Clamp(raw, .05, .95);
        var matching = (int)Math.Round(raw * members.Count);
        return new WeatherEstimateResult(probability, members.Count, matching, spec.Station,
            $"GFS ensemble {spec.Station}: {matching}/{members.Count} members match");
    }

    private async Task<Dictionary<string, double[]>?> ForecastAsync(Spec spec, CancellationToken ct)
    {
        var key = $"{spec.Lat:F4}:{spec.Lon:F4}:{spec.Date:yyyy-MM-dd}";
        if (_cache.TryGetValue(key, out var cached) && DateTimeOffset.UtcNow - cached.At < TimeSpan.FromMinutes(15))
            return cached.Data;
        var url = FormattableString.Invariant(
            $"{Endpoint}?latitude={spec.Lat}&longitude={spec.Lon}&daily=temperature_2m_max,temperature_2m_min&temperature_unit=fahrenheit&timezone=auto&start_date={spec.Date:yyyy-MM-dd}&end_date={spec.Date:yyyy-MM-dd}&models={ModelName}");
        try
        {
            using var response = await http.GetAsync(url, ct);
            response.EnsureSuccessStatusCode();
            using var json = JsonDocument.Parse(await response.Content.ReadAsStringAsync(ct));
            if (!json.RootElement.TryGetProperty("daily", out var daily) || daily.ValueKind != JsonValueKind.Object)
                return null;
            var data = new Dictionary<string, double[]>();
            foreach (var item in daily.EnumerateObject())
                if (item.Value.ValueKind == JsonValueKind.Array)
                    data[item.Name] = item.Value.EnumerateArray().Where(v => v.ValueKind == JsonValueKind.Number)
                        .Select(v => v.GetDouble()).ToArray();
            _cache[key] = (DateTimeOffset.UtcNow, data);
            return data;
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or JsonException)
        {
            return null;
        }
    }
}
